boards/arm/stm32f746g_disco/preinit.c

```diff
@@ -0,0 +1,155 @@
+	register __IO u32_t index;
+
+	/* Reset the RCC clock configuration */
+	RCC->CR = (u32_t)0x00000083;
+	RCC->CFGR = (u32_t)0x00000000;
+	RCC->PLLCFGR = (u32_t)0x24003010;
+	RCC->CIR = (u32_t)0x00000000;
+
+	/* Enable GPIO clocks for FMC pins */
+	RCC->AHB1ENR |= (
+		RCC_AHB1ENR_GPIOCEN |
+		RCC_AHB1ENR_GPIODEN |
+		RCC_AHB1ENR_GPIOEEN |
+		RCC_AHB1ENR_GPIOFEN |
+		RCC_AHB1ENR_GPIOGEN |
+		RCC_AHB1ENR_GPIOHEN
+	);
+
+	/* Cofigure the GPIO Pins for FMC */
+	configure_gpio_fmc('C', 3);
+	configure_gpio_fmc('D', 0);
+	configure_gpio_fmc('D', 1);
+	configure_gpio_fmc('D', 8);
+	configure_gpio_fmc('D', 9);
+	configure_gpio_fmc('D', 10);
+	configure_gpio_fmc('D', 14);
+	configure_gpio_fmc('D', 15);
+	configure_gpio_fmc('E', 0);
+	configure_gpio_fmc('E', 1);
+	configure_gpio_fmc('E', 7);
+	configure_gpio_fmc('E', 8);
+	configure_gpio_fmc('E', 9);
+	configure_gpio_fmc('E', 10);
+	configure_gpio_fmc('E', 11);
+	configure_gpio_fmc('E', 12);
+	configure_gpio_fmc('E', 13);
+	configure_gpio_fmc('E', 14);
+	configure_gpio_fmc('E', 15);
+	configure_gpio_fmc('F', 0);
+	configure_gpio_fmc('F', 1);
+	configure_gpio_fmc('F', 2);
+	configure_gpio_fmc('F', 3);
+	configure_gpio_fmc('F', 4);
+	configure_gpio_fmc('F', 5);
+	configure_gpio_fmc('F', 11);
+	configure_gpio_fmc('F', 12);
+	configure_gpio_fmc('F', 13);
+	configure_gpio_fmc('F', 14);
+	configure_gpio_fmc('F', 15);
+	configure_gpio_fmc('G', 0);
+	configure_gpio_fmc('G', 1);
+	configure_gpio_fmc('G', 4);
+	configure_gpio_fmc('G', 5);
+	configure_gpio_fmc('G', 8);
+	configure_gpio_fmc('G', 15);
+	configure_gpio_fmc('H', 3);
+	configure_gpio_fmc('H', 5);
+
+	/* Enable the FMC interface clock */
+	RCC->AHB3ENR |= RCC_AHB3ENR_FMCEN;
+
+	/* Configure and enable SDRAM bank1 */
+	FMC_Bank5_6->SDCR[0] = 0x00001954;
+	FMC_Bank5_6->SDTR[0] = 0x01115351;
+
+	/* SDRAM initialization sequence */
+	/* Clock enable command */
+	FMC_Bank5_6->SDCMR = 0x00000011;
+	temp_register = FMC_Bank5_6->SDSR & 0x00000020;
+	while ((temp_register != 0) && (timeout-- > 0)) {
+		temp_register = FMC_Bank5_6->SDSR & 0x00000020;
+	}
+
+	/* Delay */
+	for (index = 0; index < 1000; index++) {
+	}
+
+	/* PALL command */
+	FMC_Bank5_6->SDCMR = 0x00000012;
+	timeout = 0xFFFF;
+	while ((temp_register != 0) && (timeout-- > 0)) {
+		temp_register = FMC_Bank5_6->SDSR & 0x00000020;
+	}
+
+	/* Auto refresh command */
+	FMC_Bank5_6->SDCMR = 0x000000F3;
+	timeout = 0xFFFF;
+	while ((temp_register != 0) && (timeout-- > 0)) {
+		temp_register = FMC_Bank5_6->SDSR & 0x00000020;
+	}
+
+	/* MRD register program */
+	FMC_Bank5_6->SDCMR = 0x00044014;
+	timeout = 0xFFFF;
+	while ((temp_register != 0) && (timeout-- > 0)) {
+		temp_register = FMC_Bank5_6->SDSR & 0x00000020;
+	}
+
+	/* Set refresh count */
+	temp_register = FMC_Bank5_6->SDRTR;
+	FMC_Bank5_6->SDRTR = (temp_register | (0x0000050C << 1));
+
+	/* Disable write protection */
+	temp_register = FMC_Bank5_6->SDCR[0];
+	FMC_Bank5_6->SDCR[0] = (temp_register & 0xFFFFFDFF);
```

Comment:
This all runs before the rest of Zephyr, so it can't really be in a zephyr driver if I understand correctly. The main reason I did it this way was because it's how the problem was solved for the IMX boards.

Reply:
Even if this is a preinit driver, it should be doable to implement it as a generic STM32 driver (could potentially be under `soc/arm/st_stm32` if there is no better location) so it could be used by more boards and socs. 
Also, the pin information configuration would be the same before and after init so it would be nice to share the same configuration (eventually provided by dts).

Reply:
I'll try to sort that out.

At the dev meeting on Thursday, can you discuss the early_init code (the first commit) so I at least know if we'll use that solution.

Reply:
ok, sure

Reply:
I've been doing some tests and investigation.

It seems like it'd be too much to create a driver in zephyr/drivers/memory_controller for example...and then call that in the early_init code. It requires too much of the zephyr system.

So I've created the early_init inside of soc/arm/st_stm32/stm32f7 instead. That works fine...the next step is to get it to read the config from the dts instead. That shouldn't be a problem. I will also check to see if stm32f4 etc need this and will move it if needed.

Does this sound ok to you?

Reply:
> I've been doing some tests and investigation.
> 
> It seems like it'd be too much to create a driver in zephyr/drivers/memory_controller for example...and then call that in the early_init code. It requires too much of the zephyr system.

Ok, thanks for having had a look. I don't think it is something achievable indeed (today at least).

> So I've created the early_init inside of soc/arm/st_stm32/stm32f7 instead. That works fine...the next step is to get it to read the config from the dts instead. That shouldn't be a problem. I will also check to see if stm32f4 etc need this and will move it if needed.
> 
> Does this sound ok to you?

Sure. Having this STM32 generic would be great indeed.